File: Array/objc_based_array_sol/find_duplicate.m
#import <Foundation/Foundation.h>

NSArray *findDuplicates(NSArray *array) {
    NSMutableArray *uniqueElements = [NSMutableArray array];
    NSMutableArray *duplicates = [NSMutableArray array];

    for (NSNumber *element in array) {
        if ([uniqueElements containsObject:element]) {
            [duplicates addObject:element];
        } else {
            [uniqueElements addObject:element];
        }
    }

    return [duplicates copy];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSArray *array = @[@1, @2, @2, @3, @4, @4, @5];
        NSArray *duplicateElements = findDuplicates(array);
        
        NSLog(@"Duplicate elements in the array: %@", duplicateElements);
    }
    return 0;
}

File: Array/swift_based_array_sol/find_duplicate.swift
func findDuplicates(withArray array: [Int]) -> [Int] {
    var uniqueElements = Set<Int>()
    var duplicates = Set<Int>()

    for element in array {
        if uniqueElements.contains(element) {
            duplicates.insert(element)
        } else {
            uniqueElements.insert(element)
        }
    }

    return Array(duplicates)
}

let array = [1, 2, 2, 3, 4, 4, 5]
let duplicateElements = findDuplicates(withArray: array)
print("Duplicate elements in the array: \(duplicateElements)")